subsys/debug/gpio_frontend/gpio_frontend.c

```diff
@@ -0,0 +1,677 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <logging/log_frontend.h>
+#include <hal/nrf_gpio.h>
+
+#if (CONFIG_LOG_FRONTEND_PORT_NUMBER == 0)
+	#define LOG_FRONTEND_GPIO_REG_PTR NRF_P0
+#if CONFIG_BOARD_NRF9160_PCA10090 || CONFIG_BOARD_NRF9160_PCA10090NS
+	#define BASE_ADDRESS NRF_P0_NS_BASE
+#else
+	#define BASE_ADDRESS NRF_P0_BASE
+#endif
+#elif (CONFIG_LOG_FRONTEND_PORT_NUMBER == 1)
+	#define LOG_FRONTEND_GPIO_REG_PTR NRF_P1
+#if CONFIG_BOARD_NRF9160_PCA10090 || CONFIG_BOARD_NRF9160_PCA10090NS
+	#define BASE_ADDRESS NRF_P1_NS_BASE
+#else
+	#define BASE_ADDRESS NRF_P1_BASE
+#endif
+#endif
+
+#define MASK_OF_LOG_PINS	(0x1ff << CONFIG_LOG_FRONTEND_LSB_PIN_NUMBER)
+#define MASK_OF_CLK_PIN		BIT(CONFIG_LOG_FRONTEND_LSB_PIN_NUMBER + 8)
+#define CLEAR_ADDRESS_OFFSET	offsetof(NRF_GPIO_Type, OUTCLR)
+#define SET_ADDRESS_OFFSET	offsetof(NRF_GPIO_Type, OUTSET)
+#define HEXDUMP_STR_BYTES	0xffffffff
+
+union log_msg_ids_convert {
+	struct log_msg_ids ids;
+	u16_t raw;
+};
+
+static const u8_t sync_frame[] = {0x61, 0x62, 0x63, 0x64,
+				  0x65, 0x66, 0x67, 0x68};
+
+/*
+ * Note: Due to an emphasis on performace, this module uses assembly code
+ * to acheive as low as possible delays. When similar results are attainable
+ * with C code, this should be rewritten.
+ */
```

Comment:
It needs to be clear that this code could not have been written in C to reach the required real-time requirements. As it is written now this is not clear, it is only stated to be so.